File: python/2020/Day3.py
import sys
from common import DataAnalyzer


def traverse(course, slope=[3,1], position=[0,0]):
    width = len(course[0])
    height = len(course)
    trees = 0

    while (position[0] + 1 < height):
        position[1] += slope[0]
        position[0] += slope[1]

        if position[1] >= width:
            position[1] = position[1] - width

        if course[position[0]][position[1]] == '#':
            trees +=1

    return trees


def second():
    data = DataAnalyzer.text("2020/day3.txt")
    count = 1
    for slope in [[1, 1],[3, 1],[5, 1],[7, 1],[1, 2]]:
        count *= traverse(data, slope, [0,0])
    
    print("(3.2) trees hit is {:d}".format(count))


def first():
    data = DataAnalyzer.text("2020/day3.txt")
    print("(3.1) trees hit is {:d}".format(traverse(data)))


def solve(puzzle):
    if puzzle == '1':
        first()
    elif puzzle == '2':
        second()
    else:
        first()
        second()


if __name__ == '__main__':
    solve(sys.argv[1])

File: js/2020/Day3.js
const {DataAnalyzer} = require('../common/DataAnalyzer.js')
let data = DataAnalyzer.strs("2020/day3.txt")


function tree(data, x, y) {
    return data[x][y] === '#' ? 1 : 0;
}

function calc(x=0, y=0, i=3, j=1) {
    trees = tree(data, x, y)
    while (x + j < data.length) {
        x += j;
        y = (y + i) % data[x].length;
        trees += tree(data, x, y)
    }

    return trees;
}

function one() {
    answer = calc()
    console.log("(2020.3.1): " + answer)
}

function two() {
    answer = 1
    slopes = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
    
    slopes.forEach(slope => {
        answer *= calc(0, 0, slope[0], slope[1])
    });

    console.log("(2020.3.2): " + answer)
}

switch(process.argv.slice(2)[0]) {
    case '1':
        one();
        break;

    case '2':
        two();
        break;
    
    default:
        one();
        two();
}